### DISCUSSION TITLE
Anyone else here using sqlc over the default DAO? Here is our experience. Share yours, too.

### DISCUSSION THREAD

**Discussion by preslavrachev** [pocketbase/pocketbase#4984] [Show and tell]
For a new project of ours, we picked Pocketbase because we could make use of SQLite, and everything else that Pocketbase adds on top is just a huge time saver. However, we intentionally decided to stay away from one thing - the DAO and query builder. Nothing bad with those at all, it's just that we'd spent enough time working with GORM's and various other ORM syntaxes over time that we got sick of learning each and eve one's quirks and tweaks. Over time, we've settled on going SQL-first, with the help of a fantastic code generator like [sqlc](https://sqlc.dev/). So, we decided to give it a try in a Pocketbase setup, and it seems to work without any issues so far.

What sounded like a recipe for disaster, is actually pretty straightworward. The only downside is that PB uses Go code for its migrations, and sqlc needs a SQL schema file to generate models and all, so after ever database change, we export the entire database schema like this:

```bash
sqlite3 pb_data/data.db .schema > store/sql/schema/schema.sql
```

and keep the SQL schema version controlled. Yes, I am aware that the schema must always be in sync with the migrations state, otherwise we might run in inconsistency danger, but that's something we can potentially even automate somehow (@ganigeorgiev or is that something PB can do out of the box, without the need for 3rd-party tooling?).

The result is a much leaner querying setup - we write the queries in pure SQL, sqlc will generate the code for us, and even export a handy `Querier` interface for us, so the rest of the code can be tested without touching the DB at all. As for testing the DB integrity, it's as easy as loading that `schema.sql` file and creating an in-memory SQLite database that we can test each query over. At the end of each test, we can easily throw away the DB and recreate it again; fire and forget.

I was wondering, if anyone else is using the same combo in their PB apps, or might benefit from doing so. Let me know of any potential pitfalls we might be going into.

Cheers,
Preslav

**Comment by lumosminima** [comment]
Do [db hooks](https://pocketbase.io/docs/go-event-hooks/#db-hooks) fire when using this setup?

**Reply by preslavrachev** [reply]
Excellent question! Right now, we aren't really using the hooks since we rely on scheduled jobs to frequently poke at the database and do certain things. It's the good old polling approach that is guaranteed to work forever. However, the topic of hooks is something we might make use of, so I'll definitely check that out.

My guess is, if PB uses a polling mechanism on the DB side, they'd probably work. If it is more the DAO firing an event every time it gets to do something, then they won't. I'm betting on the former, but @ganigeorgiev can shed some more light here.

**Reply by ganigeorgiev** [reply]
@preslavrachev We don't do polling for the realtime events if that's what you are asking. The realtime events are triggered only when the `Record/Model` changes were done through the `app.Dao()`. Or in other words, after successful create/update/delete we simply broadcast the change in a fan-out like pattern to the related subscribers.

If sqlc works for your team - great, but if you are not using the PocketBase Web APIs and existing record helpers to me it seems overcomplicated and I think you may be better with dropping PocketBase and simply using SQLite directly.

_As a side-note, if you haven't already, see also https://github.com/pocketbase/pocketbase/discussions/4355._

**Comment by ashxdev** [comment]
@preslavrachev Hi, I'm so interesting in this topic, could you share a code or repository with kind of setup?

**Reply by preslavrachev** [reply]
We can't share access to the repo since it is a private project, but if this is something that people are interested in, I'd make sure to blog about it at some point. Make sure to subscribe to [my blog](https://preslav.me) to keep yourself up-to-date.

**Reply by marvincaspar** [reply]
@preslavrachev when to you plan to publish your blog post?

**Comment by tgrosinger** [comment]
I'm trying to set this up too. How are you accessing the `sql.DB` for passing to `sqlc.New()`?
